src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -243,6 +302,80 @@ func (a *StorageMinerActorCode_I) _expirePreCommittedSectors(rt Runtime) {
 
 }
 
+func (st *StorageMinerActorState_I) _updateSectorUtilization(rt Runtime) []deal.DealID {
+	// TODO: verify if we should update Sector utilization for failing sectors
+	// depends on decision around collateral requirement for inactive power
+	// and what happens when a failing sector expires
+
+	// this may not work
+	ret := deal.CompactDealSet(make([]byte, 0))
+
+	for _, sectorNo := range st.Impl().ProvingSet_.SectorsOn() {
+		utilizationInfo, found := st.SectorUtilization()[sectorNo]
+
+		if !found {
+			rt.Abort("sm._updateSectorUtilization: sector not found in SectorUtilization")
+		}
+
+		newUtilization := utilizationInfo.CurrUtilization()
+
+		currEpoch := rt.CurrEpoch()
+		firstExpirationEpoch := utilizationInfo.DealExpirationQueue().Peek().Expiration()
+		totalDealCount := len(st.Sectors()[sectorNo].SealCommitment().DealIDs())
+		newExpiredDealIDs := deal.CompactDealSet(make([]byte, totalDealCount))
+
+		if firstExpirationEpoch < currEpoch {
+			// this deal has expired
+			newExpiredDeal := utilizationInfo.DealExpirationQueue().Pop()
+			newUtilization -= newExpiredDeal.PayloadPower()
+			st.SectorUtilization()[sectorNo].Impl().ActiveDealIDs_.Remove(newExpiredDeal.DealID())
+			newExpiredDealIDs.Add(newExpiredDeal.DealID())
+
+			// this is a go while loop
+			firstExpirationEpoch = utilizationInfo.DealExpirationQueue().Peek().Expiration()
+
+		}
+
+		st.SectorUtilization()[sectorNo].Impl().CurrUtilization_ = newUtilization
+		ret = ret.Extend(newExpiredDealIDs)
+	}
+
+	return ret.DealsOn()
+
+}
```

Comment:
this function is complicated, hard to read, and doing a lot. maybe break it apart w/ a helper function or two that make the logic and intent super clear, so we can read it very simply.

Reply:
will do, this function is also where the heavy-lifting of iterating all active deals and update sector utilization happens